nerium/main.py

```diff
@@ -13,69 +14,135 @@
             return None
 
+
+class Query():
+    """ Finds query in registry, and looks up ResultSet subclass from
+    path file extension. Submits query_path to ResultSet and returns results
+    """
+
+    def __init__(self, query_name, **kwargs):
+        self.query_name = query_name
+        self.kwargs = kwargs
+
     def result_set(self):
-        """ if your functionResultSet subclass needs no script,
-            simply return self.result() here.
-        """
-        if not self.get_query_path():
-            return [
-                {
-                    'error': "No query found matching {}".format(
-                        self.query_name)
-                }
-            ]
+        registry = QueryRegistry()
+        query = registry.get_query(self.query_name)
+        if not query:
+            return [{
+                'error':
+                "No query found matching {}".format(self.query_name)
+            }]
         else:
-            return self.result()
+            query_path = query['query_path']
+            result_mods = import_module('nerium.contrib.resultset')
+            result_cls = getattr(result_mods, query['result_cls'])
+            loader = result_cls(query_path, **self.kwargs)
+            query_result = loader.result()
+            return query_result
+
+
+class ResultSet(ABC):
+    """Generic result set class template.
+
+    Loads data source and fetches results by submitting query to data source
+    with expected parameters.
+
+
+    Args:
+        query_path (path-like object): Path to query file
+        kwargs: Parameter keys and values to bind to query
+
+    Usage:
+        Subclass and provide a 'result' method
+    """
+
+    def __init__(self, query_path, **kwargs):
+        self.query_path = query_path
+        self.kwargs = kwargs
 
     @abc.abstractmethod
     def result(self):
-        """ Return query results as a JSON-serializable Python structure
+        """ Given a query_path, return results as a serializable Python structure
         (generally a list of dictionaries)
         """
         return
 
 
-class ResultFormatter(ABC):
-    def __init__(self, result):
+class ResultFormat(ABC):
+    def __init__(self, result, format_):
         self.result = result
+        self.format_ = format_
 
     def formatted_results(self):
+        format_lookup = {
+            'affix': 'AffixFormatter',
+            'compact': 'CompactFormatter',
+            'csv': 'CsvFormatter',
+            'default': 'DefaultFormatter',
+        }
         if 'error' in self.result[0].keys():
             return self.result, 400
         else:
-            return self.format_results()
+            format_cls_name = format_lookup[self.format_]
+            format_mods = import_module('nerium.contrib.formatter')
+            format_cls = getattr(format_mods, format_cls_name)
+            return format_cls(self.result).format_results()
+
+
+class ResultFormatter(ABC):
+    def __init__(self, result):
```

Comment:
and I guess if we returned more metadata with the results, that could always need to be passed onto ResultFormatter

Reply:
At the moment, I'm thinking the frontmatter and other metadata will be handled by separate class(es)/methods from the resultset itself but I don't consider that settled until I try it.